src/backend/utils/resgroup/resgroup.c

```diff
@@ -4766,6 +4816,81 @@ moveQueryCheck(int sessionId, Oid groupId)
 	cdbdisp_clearCdbPgResults(&cdb_pgresults);
 }
 
+/*
+ * Try to give away all slot control to target process.
+ */
+static void
+resGroupGiveSlotAway(int sessionId, ResGroupSlotData **slot, Oid groupId)
+{
+	long timeout;
+	int64 curTime;
+	int64 waitStart;
+	int latchRes;
+	bool clean = false;
+	bool res = false;
+	
+	SIMPLE_FAULT_INJECTOR("resource_group_give_away1");
+
+	if (!ResGroupMoveSignalTarget(sessionId, *slot, groupId))
+		elog(ERROR, "can't send signal to process");
+	
+	waitStart = GetCurrentIntegerTimestamp();
+
+	for (;;)
+	{
+		CHECK_FOR_INTERRUPTS();
+
+		curTime = GetCurrentIntegerTimestamp();
+		timeout = gp_resource_group_move_timeout - (curTime - waitStart) / 1000;
+		if (timeout > 0)
+		{
+			PG_TRY();
+			{		
+				SIMPLE_FAULT_INJECTOR("resource_group_give_away2-1");
+				SIMPLE_FAULT_INJECTOR("resource_group_give_away2-2");
+				
+				latchRes = WaitLatch(&MyProc->procLatch,
+					WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH, timeout);
+
+				if (latchRes & WL_POSTMASTER_DEATH)
+					elog(ERROR,
+						"got WL_POSTMASTER_DEATH waiting on latch; exiting...");
+			}
+			PG_CATCH();
```

Comment:
> Which kind of errors do you expect here and in the test case? Do you expect WaitLatch ERRORs? Am I right? 
Yes.
> The logic is ok, but wrapping WaitLatch in try/catch isn't distributed spreadly in gpdb code base.
Probably, because there nothing shared to free in another cases.
Here we need to clean all flags and get the actual result of handling to decide can we free slot or not.